storm-core/src/jvm/org/apache/storm/bolt/JoinBolt.java

```diff
@@ -0,0 +1,428 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.bolt;
+
+
+import org.apache.storm.task.OutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.topology.OutputFieldsDeclarer;
+import org.apache.storm.topology.base.BaseWindowedBolt;
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.windowing.TupleWindow;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class JoinBolt extends BaseWindowedBolt {
+
+    private OutputCollector collector;
+
+    // Map[StreamName -> Map[Key -> List<Tuple>]  ]
+    HashMap<String, HashMap<Object, ArrayList<Tuple> >> hashedInputs = new HashMap<>(); // holds remaining streams
+
+    // Map[StreamName -> JoinInfo]
+    protected LinkedHashMap<String, JoinInfo> joinCriteria = new LinkedHashMap<>();
+    protected String[][] outputKeys;  // specified via bolt.select() ... used in declaring Output fields
+    protected String[] dotSeparatedOutputKeyNames; // flattened (de nested) keyNames, used for naming output fields
+    protected String outputStreamName;
+
+    // Use streamId, source component name OR field in tuple to distinguish incoming tuple streams
+    public enum Selector { STREAM, SOURCE }
+    protected final Selector selectorType;
+
+
+    /**
+     * StreamId to start the join with. Equivalent SQL ...
+     *       select .... from streamId ...
+     * @param type Specifies whether 'streamId' refers to stream name/source component
+     * @param streamId name of stream/source component
+     * @param key the fieldName to use as key for the stream (used for performing joins)
+     */
+    public JoinBolt(Selector type, String streamId, String key) {
+        selectorType = type;
+        joinCriteria.put(streamId, new JoinInfo(key) );
+    }
+
+    /**
+     * Defines the name of the output stream
+     */
+    public JoinBolt withOutputStream(String streamName) {
+        this.outputStreamName = streamName;
+        return this;
+    }
+
+    /**
+     * Performs inner Join.
+     *  SQL    :   from priorStream inner join newStream on newStream.key = priorStream.key1
+     *  same as:   new WindowedQueryBolt(priorStream,key1). join(newStream, key, priorStream);
+     *
+     *  Note: priorStream must be previously joined.
+     *    Valid ex:    new WindowedQueryBolt(s1,k1). join(s2,k2, s1). join(s3,k3, s2);
```

Comment:
Discussed this offline with @roshannaik , this is necessary when we have left outer join. Consider a join of tables T1, T2, T3. If join of T1 and T2 is left outer join it can lead to intermediate table which looks like:

T1.K1| T2.K1
-----|-------
 Y     | NULL 
Now consider the case, when we have to do inner join with T3.K1 (with value Y) the resulting table might change depending on whether it is join on T1.K1 or T2.K1